database/src/ops/tx.rs

```diff
@@ -1,64 +1,177 @@
 //! Transactions.
 
 //---------------------------------------------------------------------------------------------------- Import
+use cuprate_types::{OutputOnChain, TransactionVerificationData, VerifiedBlockInformation};
+use monero_pruning::PruningSeed;
+use monero_serai::transaction::{Timelock, Transaction};
 
-//---------------------------------------------------------------------------------------------------- Free Functions
-/// TODO
-pub fn add_transaction() {
-    todo!()
-}
+use crate::{
+    database::{DatabaseIter, DatabaseRo, DatabaseRw},
+    env::EnvInner,
+    error::RuntimeError,
+    ops::macros::{doc_add_block_inner_invariant, doc_error},
+    tables::{
+        BlockBlobs, BlockHeights, BlockInfoV1s, BlockInfoV2s, BlockInfoV3s, KeyImages, NumOutputs,
+        Outputs, PrunableHashes, PrunableTxBlobs, PrunedTxBlobs, RctOutputs, Tables, TablesMut,
+        TxHeights, TxIds, TxUnlockTime,
+    },
+    transaction::{TxRo, TxRw},
+    types::{
+        BlockHash, BlockHeight, BlockInfoLatest, BlockInfoV1, BlockInfoV2, BlockInfoV3, KeyImage,
+        Output, PreRctOutputId, RctOutput, TxHash, TxId,
+    },
+};
 
-/// TODO
-pub fn add_transaction_data() {
-    todo!()
-}
+use super::property::get_blockchain_pruning_seed;
 
+//---------------------------------------------------------------------------------------------------- Private
 /// TODO
-pub fn remove_transaction() {
-    todo!()
-}
+///
+/// TODO: document this add to the latest block height.
+///
+#[doc = doc_add_block_inner_invariant!()]
+#[doc = doc_error!()]
+///
+/// # Example
+/// ```rust
+/// # use cuprate_database::{*, tables::*, ops::block::*, ops::tx::*};
+/// // TODO
+/// ```
+#[inline]
+#[allow(clippy::needless_pass_by_ref_mut)] // TODO: remove me
+pub fn add_tx(
+    tx: &Transaction,
+    table_block_heights: &impl DatabaseRo<BlockHeights>,
+    table_tx_ids: &mut impl DatabaseRw<TxIds>,
+    table_tx_heights: &mut impl DatabaseRw<TxHeights>,
+    table_tx_unlock_time: &mut impl DatabaseRw<TxUnlockTime>,
+    table_prunable_hashes: &mut impl DatabaseRw<PrunableHashes>,
+    table_prunable_tx_blobs: &mut impl DatabaseRw<PrunableTxBlobs>,
+) -> Result<TxId, RuntimeError> {
+    let tx_id = get_num_tx(table_tx_ids)?;
+    let block_height = crate::ops::blockchain::height(table_block_heights)?;
 
-/// TODO
-pub fn remove_transaction_data() {
-    todo!()
-}
+    table_tx_ids.put(&tx.hash(), &tx_id)?;
+    table_tx_heights.put(&tx_id, &block_height)?;
 
-/// TODO
-pub fn remove_tx_outputs() {
-    todo!()
-}
+    // TODO: What exactly is a `UnlockTime (u64)` in Cuprate's case?
+    // What should we be storing? How?
+    match tx.prefix.timelock {
+        Timelock::None => (),
+        Timelock::Block(height) => todo!(), // Calculate from height?
+        Timelock::Time(time) => todo!(),
+    };
```

Comment:
How are we storing the timelock? This is inside `add_tx()`.

Reply:
just store it as a raw u64, but only store it if it is not `None`

Reply:
Are these not 2 different units of time? We don't need to treat them differently?

This is okay?
```rust
match tx.prefix.timelock {
    Timelock::None => (),
    Timelock::Block(height) => tables.tx_unlock_time_mut().put(&tx_id, &(height as u64))?,
    Timelock::Time(time) => tables.tx_unlock_time_mut().put(&tx_id, &time)?,
}
```

Reply:
> Are these not 2 different units of time? We don't need to treat them differently?

height is any value less than 500_000_000 and timestamp is any value above so we will be able to tell them apart by that